lib/sources.nix

```diff
@@ -229,6 +229,42 @@ let
 
   canCleanSource = src: src ? _isLibCleanSourceWith || !(pathHasContext (toString src));
 
+  /*
+    Get all files (recursively) that pass a filter function. Unlike other source
+    filtering functions, this will not include empty directories in the final
+    derivation.
+
+    Type: sourceLike -> (String -> Bool) -> Source
+
+    Example:
+      filterFiles ./. (lib.hasSuffix ".html")
+  */
+  filterFiles = src: filter:
+    let
+      includedPaths =
+        builtins.filter
+          filter
+          (lib.filesystem.listFilesRecursive src);
```

Comment:
This does not actually support all `sourceLike` values, such as the result of `cleanSource`. Types are a bit vague in Nix documentation, so I really can't blame you for not picking up on this subtlety.

Reply:
I just tried
```
lib.filesystem.listFilesRecursive (lib.cleanSource ./.)
```
in nixpkgs and it worked.